20221103-endpoint-authz.md

```diff
@@ -0,0 +1,304 @@
+3. If there is a global config:
+   1. If either the scope or the subject exist in the global config, then return `nil`.
+4. Return an authorization error.
+
+# Migration
+[migration]: #migration
+
+No migration is needed; support for endpoint authorization will be completely opt-in.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* If the implementation is simple, it will not support all authorization use-cases (i.e. combinations of allow and deny).
+* If the implementation allows for more complex authorization schemes, it will be more difficult to configure and more prone to implementation bugs.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## Tweaks to this design
+There are a couple of additions/deletions to this design that make it easier to define certain authorization schemas at the expense of additional complexity.
+
+### Don't allow global scopes
+Originally, this design did not include the global scope configs. It does add a small amount of complexity, but I think that usage will be common enough that including it as an option is worth it. We could require explicit authorizaiton on every endpoint instead.
+
+### Read/write endpoint classification
+It may be a reasonable assumption that most clients may want to create distinct restrictions on read and write endpoints. We could classify each endpoint as one of the two and expose a config option for each, again allowing the same variety of subject authorization definitions.
+
+```yaml
+authn:
+  method: oidc
+  oidc:
+    issuer: https://my-cool-oauth-server.com
+    audience: my-audience
+    read_authz:
+      subjects:
+        - client-id-a
+      scopes:
+        - openfga:read
+    write_authz:
+      subjects:
+        - client-id-a
+      scopes:
+        - openfga:write
+```
+
+I believe that this approach is too prescriptive; we should leave this in the hands of the operators even if it requires a bit more verbosity in the configuration.
+
+## Other approaches
+### OpenFGA-driven authorization
+Given that OpenFGA is an authorization system, we could prescribe an authorization model within OpenFGA that models access to the various OpenFGA API operations. Something along the lines of:
+```
+type openfga_api
+  relations
+    define full_access as self
+    define can_create_store as self or full_access
+    define can_write_tuple as self or full_access
+    ...etc
+```
+
+The OpenFGA authorization code would then make a `Check` call on each request to authorize a client. While dogfooding is appealing at its surface, this approach has some drawbacks:
+- It requires setup outside of just writing a configuration file.
+- It couples the authorization system to other parts of OpenFGA.
+- There is a chicken-and-egg problem; you cannot protect the system prior to starting it up.
+- OpenFGA would be prescribing an authorization model that might not work for the operator
```

Comment:
Just a note on this - the primary proposal mentioned herein is also very prescriptive (scopes/preshared keys per endpoint). A benefit of this alternative approach is that the model can be more fine grained and dynamic instead of statically prescribed. For example, with this approach the operators could potentially define their system access model using OpenFGA modeling.